lib/ansible/plugins/connection/aws_ssm.py

```diff
@@ -0,0 +1,337 @@
+# Based on the ssh connection plugin by Michael DeHaan
+#
+# Copyright: (c) 2018, Pat Sharkey <psharkey@cleo.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = """
+author: psharkey@cleo.com
+connection: aws_ssm
+short_description: execute via AWS Systems Manager
+description:
+- This connection plugin allows ansible to execute tasks on an EC2 instance via the aws ssm CLI.
+version_added: "2.8"
+requirements:
+- The control machine and the remote EC2 instance must have the aws CLI installed.
```

Comment:
Have you considered using boto3? It would allow you to use [AWSRetry](https://github.com/ansible/ansible/blob/stable-2.7/lib/ansible/module_utils/ec2.py#L78) and make [custom waiters](https://github.com/ansible/ansible/blob/stable-2.7/lib/ansible/module_utils/aws/waiters.py) to help make this more stable to use.

Reply:
I had but would need some documentation for the WebSocket connection from [start_session](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ssm.html#SSM.Client.start_session) . I think it would require some reverse-engineering on the aws cli (https://github.com/aws/amazon-ssm-agent/tree/master/agent/session/plugins).

Reply:
@s-hertel The dependency on the AWS CLI  for establishing the [terminal session](https://github.com/aws/amazon-ssm-agent/tree/master/agent/session/plugins) can be replaced using boto3 similar to [how the AWS CLI already does this](https://github.com/aws/aws-cli/blob/develop/awscli/customizations/sessionmanager.py#L60-L102). The dependency on the [plugin](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-install-plugin.html) which has the contract needed to do the bulk of the terminal session work would still be necessary. The [plugin](https://github.com/aws/amazon-ssm-agent/tree/master/agent/session/plugins) will abstract the target OS and hopefully support managed (on-premise) instances soon. Do you think this is an acceptable solution?

Reply:
https://github.com/psharkey/ansible/commit/4989a94d0e8ee6c7d8f78eca9421663c9544219c shows what it would take to switch to boto3 for the terminal session. This may open the door to pipelining these connections.

Reply:
https://github.com/ansible/ansible/pull/49652/commits/d896a119318d96eeef6577ba246cccade8d02bbd adds the switch to boto3 for the terminal connection in this PR. If sessions can be reused by just tracking the [response](https://github.com/ansible/ansible/blob/d896a119318d96eeef6577ba246cccade8d02bbd/lib/ansible/plugins/connection/aws_ssm.py#L170) with the `instance_id` there could be a performance improvement.

Reply:
@s-hertel The dependency on the AWS CLI has been removed on both sides and https://github.com/ansible/ansible/pull/49652/commits/5b02fc8f3d0f1c3588a1c9592973d152dd1469dc adds support for connections to Windows remote EC2 instances.